src/mlpack/methods/reinforcement_learning/sac_impl.hpp

```diff
@@ -0,0 +1,343 @@
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_RL_SAC_IMPL_HPP
+#define MLPACK_METHODS_RL_SAC_IMPL_HPP
+
+#include <mlpack/prereqs.hpp>
+
+#include "sac.hpp"
+
+namespace mlpack {
+namespace rl {
+
+template <
+  typename EnvironmentType,
+  typename QNetworkType,
+  typename PolicyNetworkType,
+  typename UpdaterType,
+  typename ReplayType
+>
+SAC<
+  EnvironmentType,
+  QNetworkType,
+  PolicyNetworkType,
+  UpdaterType,
+  ReplayType
+>::SAC(TrainingConfig& config,
+       QNetworkType& learningQ1Network,
+       PolicyNetworkType& policyNetwork,
+       ReplayType& replayMethod,
+       UpdaterType qNetworkUpdater,
+       UpdaterType policyNetworkUpdater,
+       EnvironmentType environment):
+  config(config),
+  learningQ1Network(learningQ1Network),
+  policyNetwork(policyNetwork),
+  replayMethod(replayMethod),
+  qNetworkUpdater(std::move(qNetworkUpdater)),
+  #if ENS_VERSION_MAJOR >= 2
+  qNetworkUpdatePolicy(NULL),
+  #endif
+  policyNetworkUpdater(std::move(policyNetworkUpdater)),
+  #if ENS_VERSION_MAJOR >= 2
+  policyNetworkUpdatePolicy(NULL),
+  #endif
+  environment(std::move(environment)),
+  totalSteps(0),
+  deterministic(false)
+{
+  // Set up q-learning and policy networks.
+  learningQ2Network = learningQ1Network;
+  targetQ1Network = learningQ1Network;
+  targetQ2Network = learningQ2Network;
+  if (learningQ1Network.Parameters().is_empty())
+    learningQ1Network.ResetParameters();
+  if (learningQ2Network.Parameters().is_empty())
+    learningQ2Network.ResetParameters();
+  if (targetQ1Network.Parameters().is_empty())
+    targetQ1Network.ResetParameters();
+  if (targetQ2Network.Parameters().is_empty())
+    targetQ2Network.ResetParameters();
+  if (policyNetwork.Parameters().is_empty())
+    policyNetwork.ResetParameters();
+
+  #if ENS_VERSION_MAJOR == 1
+  this->qNetworkUpdater.Initialize(learningQ1Network.Parameters().n_rows,
+                                   learningQ1Network.Parameters().n_cols);
+  #else
+  this->qNetworkUpdatePolicy = new typename UpdaterType::template
+      Policy<arma::mat, arma::mat>(this->qNetworkUpdater,
+                                   learningQ1Network.Parameters().n_rows,
+                                   learningQ1Network.Parameters().n_cols);
+  #endif
+
+  #if ENS_VERSION_MAJOR == 1
+  this->policyNetworkUpdater.Initialize(policyNetwork.Parameters().n_rows,
+                                        policyNetwork.Parameters().n_cols);
+  #else
+  this->policyNetworkUpdatePolicy = new typename UpdaterType::template
+      Policy<arma::mat, arma::mat>(this->policyNetworkUpdater,
+                                   policyNetwork.Parameters().n_rows,
+                                   policyNetwork.Parameters().n_cols);
+  #endif
+
+  targetQ1Network.Parameters() = learningQ1Network.Parameters();
```

Comment:
We might want to do the same thing for the other networks as well.

Reply:
Yeah, I'd like to do that. But https://github.com/mlpack/mlpack/pull/2487#issuecomment-665736751 is still a problem. The fact that the two implementations behave differently inspite of them having same hyperparameters. That's why, I'm currently trying to run both of them on mlpack's Pendulum env, and set a fixed seed for both, and then compare outputs, so as to find at what point, the two of them differ.

Reply:
Ahh, found the problem! 

Actually implementations https://gist.github.com/nishantkr18/56601556a0f46ad161f4f4baaff46377 and https://gist.github.com/nishantkr18/29f9a58bc479eab8e9cfea9e38785b51 mentioned in https://github.com/mlpack/mlpack/pull/2487#issuecomment-665736751 are exactly the same, accept at one point : [this line](https://gist.github.com/nishantkr18/29f9a58bc479eab8e9cfea9e38785b51#file-sac_full-cpp-L111). Notice that here, we always store `false` (in place of a boolean indicating end of episode) at the replay buffer.

This was the only reason for [this](https://gist.github.com/nishantkr18/29f9a58bc479eab8e9cfea9e38785b51) implementation to perform so well.

Reply:
Nice catch.